pinot-spi/src/main/java/org/apache/pinot/spi/utils/ByteArray.java

```diff
@@ -54,6 +53,9 @@ public class ByteArray implements Comparable<ByteArray>, Serializable {
 
   private final byte[] _bytes;
 
+  // Hash for empty ByteArray is 1
+  private int _hash = 1;
+
   public ByteArray(byte[] bytes) {
```

Comment:
Java’s string added a flag whether the hash had been computed to avoid computing the hash every time if the hash code happened to be 0, the same should be done here in case the hash happens to be 1 (finding {x_i} such that sum(x_i * 31^(n-i)) = 1 gives the byte arrays which collide, it’s easy to construct examples and they do occur in reality)

Reply:
@xiangfu0 Good point. We don't reuse the `byte[]` in the `ByteArray` right now, but there is no way to enforce that without cloning a byte array during construction, which will add overhead.
Added some comments to the javadoc